sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala

```diff
@@ -121,7 +120,7 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry)
    */
   private[sql] def registerJavaUDAF(name: String, className: String): Unit = {
     try {
-      val clazz = Utils.classForName[AnyRef](className)
+      val clazz = session.artifactManager.classloader.loadClass(className)
```

Comment:
One follow-up here would be to cache the ArtifactManager classloader. I think we create that thing over and over.

Reply:
Changed it to a `lazy val`. It is safe because during initialisation it only reads read-only SparkEnv configs

Reply:
Agree. We can invalid the cache when a new JAR is added.

Reply:
Created https://issues.apache.org/jira/browse/SPARK-50243.